### PR TITLE
Instrumenting Polycube with Prometheus Metrics

### PR DESCRIPTION
**Pull request by pinoOgni**
Here we go again. 

So far, it’s hard to inspect the status of running cubes to detect anomalies.

With this PR, we introduce the possibility to expose cubes metrics via an Open Metrics endpoint. 
This enables to monitor and collect statistics how they work at runtime, without having to modify the code of the cubes. 

Further Design Principles:

- Use a standard format to foster integration with existing tools

- Minimize the code to be maintained

### CODE REVIEW

**Comment by FedeParola** [comment]
Hi @pinoOgni, would it possible to avoid having a whole copy of the sources of the libraries you used?
For example for `prometheus-cpp` we could just clone and install it upon Polycube installation, the same way it is done for `pistache` and other libraries: https://github.com/polycube-network/polycube/blob/6437c4d1e585e04498a19af8162b1e4f7a194528/scripts/pre-requirements.sh#L65-L79
For `jsoncons` maybe a git submodule would help keep code more organized, or we could just copy the content of include folders as suggested in the README of the project.
Let me know what you think about it.

**Comment by pinoOgni** [comment]
Hi @FedeParola,  about the libraries, I spoke with @frisso  and it had been decided not to add submodules and delete everything unnecessary. 

1. for jsoncons I followed the readme by putting "jsoncons" and "jsonpath" which is an extension that I used to extract the values   from the json of the cubes. The only extra thing I did was to add the cmake, CMakeLists, acknowledgments and license. It would be possible to remove some files inside "jsoncons" and "jsonpath" which are not needed but I left them for completeness and for future changes. 

2. for prometheus-cpp, at the beginning I had tried to do as for Pistache only that even in this case I would have brought two submodules that I didn't need and also two parts of prometheus-cpp that is "pull" and "push  ".  The alternative could be to clone, remove what is not needed and modify the cmakelists but it doesn't seem to me the best choice.  

Let me know what you think, I'm open to anything.

File: src/polycubed/src/rest_server.h [hunk 1]
```diff
@@ -153,6 +164,29 @@ class RestServer {
   static std::string blacklist_cert_path;
 
   std::shared_ptr<spdlog::logger> logger;
+
+  // the struct Metric for now is composed of two vectors for each metric (Gauge
+  // and Counter), one indicates the family of the metric and one the metric
+  // itself Note: families combine values with the same name, but distinct label dimensions
+  // TODO maybe there are best ways
+  struct Metric {
```
[the remark this answers is not in the document]
Comment (pinoOgni) [reply src/polycubed/src/rest_server.h:172]: Hi @FedeParola , regarding the first part I think it is better to keep the two functions separate. The ```RestServer::create_metrics()``` function reads the data deriving from the parsing of the yangs and creates the data structures. ```RestServer::get_metrics()``` uses these structures only to add the values ​​I get with jsoncons. I don't delete any cube, I just do checks:
1. does the cube exist and do I have the metric? Then I update the value
2. does the cube exist and I don't have the metric? then I add a label that is cubeName = "cubeName"
3. do I have the metric of a cube that has been cleared? then I have to clear all his metrics. This is because if you create another cube of the same service with the same name on a logical level, they are two different cubes and therefore two different stories, but if you do not delete it, it would result as if it were the same cube.

Another reason why I thought about dividing the two parts is that Prometheus does the endpoint scraping which is assigned every 15 seconds so I save time by not recreating the data structures.

Tell me though if I misunderstood something from your comment :muscle:

File: src/services/pcn-router/datamodel/router.yang [hunk 2]
```diff
@@ -23,11 +23,20 @@ module router {
...
+        polycube-base:help-metric "Number of ports of the router with ip";
+        polycube-base:type-operation "FILTER";
       }
 
       list secondaryip {
         description "Additional IP addresses for the port";
         key "ip";
+        polycube-base:name-metric "router_secondaryip_ports";
+        polycube-base:type-metric "G";
+        polycube-base:path-metric '$.ports[0].secondaryip.length';
```
[the remark this answers is not in the document]
Comment (pinoOgni) [reply src/services/pcn-router/datamodel/router.yang:38]: Thanks, nice observation! (I made many changes over time and I was sure that something had escaped me. After the exams I will also check the other yangs)

This metric should be something like "number of secondary ip addresses of the router" with path equal to "$ .ports [*]. secondaryip.length" and also with operation type equal to "FILTER".

Maybe we could extend the metric as "number of secondary ip addresses of port X" but right now I know that it would take another label, so something like {cubeName = "r1", portsName = "p1"} = value. 

However, this would mean changing both the yang and the server because it is a double level of label. For now I would opt for the first solution, there will be changes in the future. I talk both at the level of adding metrics in the yang (in fact not all services have metrics) and at the level of creating metrics.

A "dratic"  choice that comes to me at the moment (I don't know how convenient) would be to make sure that there is all the information in the json of a cube to be able to create the metrics at run-time and then also add the value. This should also answer the doubt about the two functions on the server.

Let me know what you think and thanks a lot again for the observation @FedeParola :+1:

File: src/polycubed/src/rest_server.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (FedeParola) [reply src/polycubed/src/rest_server.h:172]: I understand, thanks for your clarification. What about `struct Metric`? Do you think it would be possible to simplify it to only include two maps? I mean, now you are using a vector to store the metrics and a map to perform the mapping MetricName -> PositionInTheVector, but why not storing directly the metric in a map indexed by the MetricName?

File: src/services/pcn-router/datamodel/router.yang (hunk 2 above)
[the remark this answers is not in the document]
Comment (FedeParola) [reply src/services/pcn-router/datamodel/router.yang:38]: > Thanks, nice observation! (I made many changes over time and I was sure that something had escaped me. After the exams I will also check the other yangs)
> 
> This metric should be something like "number of secondary ip addresses of the router" with path equal to "$ .ports [*]. secondaryip.length" and also with operation type equal to "FILTER".
 
I think this solution would be the best for now. Unfortunately JSONPath is not very flexible, and complex metrics require processing also on the server side (as the need of a `type-operation` field shows)

**Comment by FedeParola** [comment]
> Hi @FedeParola, about the libraries, I spoke with @frisso and it had been decided not to add submodules and delete everything unnecessary.
> 
>     1. for jsoncons I followed the readme by putting "jsoncons" and "jsonpath" which is an extension that I used to extract the values   from the json of the cubes. The only extra thing I did was to add the cmake, CMakeLists, acknowledgments and license. It would be possible to remove some files inside "jsoncons" and "jsonpath" which are not needed but I left them for completeness and for future changes.
> 
>     2. for prometheus-cpp, at the beginning I had tried to do as for Pistache only that even in this case I would have brought two submodules that I didn't need and also two parts of prometheus-cpp that is "pull" and "push  ".  The alternative could be to clone, remove what is not needed and modify the cmakelists but it doesn't seem to me the best choice.
> 
> 
> Let me know what you think, I'm open to anything.

The reason of this approach is to reduce the size of Polycube installation? Or is there anything else?

File: src/services/pcn-router/datamodel/router.yang (hunk 2 above)
[the remark this answers is not in the document]
Comment (pinoOgni) [reply src/services/pcn-router/datamodel/router.yang:38]: OK thanks. I'll change that metric and update the PR.

About JSONPath you're absolutely right. At the beginning I was undecided on what to use, I found jsoncons and I got to work, only towards the end I started to understand its limitations. Maybe in the future it could be changed with some other library, like, I don't know, jq could be an alternative (It shouldn't be difficult to change the library, it's just a matter of changing the "path" in the metrics and something else)

**Comment by pinoOgni** [comment]
@FedeParola  
Well I think it's to reduce the size but also not to carry parts that are not needed. For example in prometheus-cpp, removing "pull" and "push" I could also remove two submodules, google-test and civit-web.

File: src/polycubed/src/rest_server.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (pinoOgni) [reply src/polycubed/src/rest_server.h:172]: Hi @FedeParola , as far as struct Metric is concerned. It is made up of two vectors and two maps simply because I started following the examples of Prometheus and in the meantime I was adapting the code to my needs. Yesterday I had some time to think about it and I modified the struct and also the rest server. This is the struct of now:

```
  struct Metric {
    std::map<std::string, std::reference_wrapper<prometheus::Family<prometheus::Counter>>> counters_map;
    std::map<std::string, std::reference_wrapper<prometheus::Family<prometheus::Gauge>>> gauges_map;
  };
```

In the rest server in terms of performance, not much has improved since I still have to cycle through all the metrics of each cube, but in doing so I saved something in terms of data structures, so thanks again for your observation :muscle: .

File: src/services/pcn-router/datamodel/router.yang (hunk 2 above)
[the remark this answers is not in the document]
Comment (pinoOgni) [reply src/services/pcn-router/datamodel/router.yang:38]: Hi @FedeParola . About my last osservation, I was wrong. I had some time to think about it and the correct solution is this:

```
polycube-base: path-metric 'sum ($. ports [*]. secondaryip.length)';
```
Without any "FILTER" as a type of operation. This is because, in this case I need the sum of the values ​​of the vector returned by "length".

File: src/polycubed/src/rest_server.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (FedeParola) [reply src/polycubed/src/rest_server.h:172]: Great! I know there isn't a performance improvement but this change should make the code easier to understand.

File: src/services/pcn-router/datamodel/router.yang (hunk 2 above)
[the remark this answers is not in the document]
Comment (FedeParola) [reply src/services/pcn-router/datamodel/router.yang:38]: That's great, I didn't notice it but jsoncons extends the way aggregation functions can be used (the expression you wrote shouldn't work with the "standard" JSONPath). These aggregation functions should remove the limitations of JSONPath and it should be possible to completely drop the `type-operation` extension. For example here https://github.com/polycube-network/polycube/blob/8668f26b61b79678a0d64f5ea6a3b6ded008b518/src/services/pcn-router/datamodel/router.yang#L25-L30 we could just use the expression `count($.ports[?(@.ip != '')])` and remove the "FILTER", correct me if I'm wrong.

[the remark this answers is not in the document]
Comment (pinoOgni) [reply src/services/pcn-router/datamodel/router.yang:38]: I thought about it too, but the problem arises when filters and functions are put together. 

For example, it is not possible to apply "length" to a query that contains a filter and neither can ``` count($.ports[?(@.ip != '')]) ```. As far as I understand it is a lack of JSONPath. The way I have found to "fix" this problem is to use type-operation. Another possible way is to make two queries and in the second one give as input the output of the first query.

[the remark this answers is not in the document]
Comment (FedeParola) [reply src/services/pcn-router/datamodel/router.yang:38]: You are right, functions don't support filter expressions, this is really unfortunate, it would have solved a lot of problems.

File: src/libs/README.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+Jsoncons
```
Comment (frisso) [inline src/libs/README.md:1]: ADD:
Libraries used by Polycube
==================

This folder includes all the libraries that are required by Polycube and that are statically included in the source files.

Although this makes more difficult the update of the above libraries to more recent versions, these are used for limited purposes and hence this necessity should not be so common. In exchange, this guarantees that the Polycube code always compiles with the above components.

File: src/libs/README.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+--------
```
Comment (frisso) [inline src/libs/README.md:2]: Add if the library is entirely copied here, or you pruned the unnecessary files.
Make sure that people know how to proceed in case they want to update the library (e.g., in case of security bugs are discovered, so we need to update the code).

File: src/libs/README.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+The library chosen to obtain the value of a metric from the json of cube is ` Jsoncons <https://github.com/danielaparker/jsoncons>`_ . It allows to use JsonPath.
+
+
+
+Prometheus-cpp
+---------------
+The library chosen to create metrics was `Prometheus-cpp <https://github.com/jupp0r/prometheus-cpp.git>`_ .
```
Comment (frisso) [inline src/libs/README.md:9]: If possible, add a brief description of the other libraries contained in this folder.

File: src/libs/jsoncons/build_files/cmake/Linux.cmake~ [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#
```
Comment (frisso) [inline src/libs/jsoncons/build_files/cmake/Linux.cmake~:1]: Is this file needed?

File: src/libs/jsoncons/build_files/cmake/config.cmake.in [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# jsoncons cmake module
```
Comment (frisso) [inline src/libs/jsoncons/build_files/cmake/config.cmake.in:1]: Is this file needed? If we have the CMakeList, other makefiles should not be needed, as they are generated automatically.

File: src/libs/README.md [hunk 8]
```diff
@@ -0,0 +1,9 @@
+Jsoncons
+--------
+The library chosen to obtain the value of a metric from the json of cube is ` Jsoncons <https://github.com/danielaparker/jsoncons>`_ . It allows to use JsonPath.
```
Comment (frisso) [inline src/libs/README.md:3]: I would also add how to compile this library.
In case this makefile is called automatically from the main makefile, please say that explicitely, so that the user doesn't have to bother about how to compile this code.

File: src/libs/prometheus-cpp/cmake/prometheus-cpp-config.cmake.in [hunk 9]
```diff
@@ -0,0 +1,11 @@
+@PACKAGE_INIT@
```
Comment (frisso) [inline src/libs/prometheus-cpp/cmake/prometheus-cpp-config.cmake.in:1]: Is this file needed?

File: src/polycubed/src/CMakeLists.txt [hunk 10]
```diff
@@ -76,6 +78,7 @@ set(polycubed_libraries
   viface
   dl
   pistache
+  prometheus-cpp-core
```
Comment (frisso) [inline src/polycubed/src/CMakeLists.txt:81]: So, prometheus-cpp-core will be compiled as a library and linked statically, isn't it? I would write it in the main readme (the one that summarizes the external libs used).
Instead, it seems that jsoncons is compiled within Polycube, isn't it? I would write it in the main readme as well.

File: src/polycubed/src/cube_factory_impl.cpp [hunk 11]
```diff
@@ -193,5 +193,13 @@ std::vector<std::shared_ptr<BaseCubeIface>> CubeFactoryImpl::get_cubes() {
   return r;
 }
 
+std::vector<std::string> CubeFactoryImpl::get_names_cubes() const {
```
Comment (frisso) [inline src/polycubed/src/cube_factory_impl.cpp:196]: I know that there are not many comments in Polycube. However, it would be nice to add something more in order to improve the code.
So, for instance, what does this function?

File: src/polycubed/src/polycubed_core.cpp [hunk 12]
```diff
@@ -636,5 +636,36 @@ CubesDump *PolycubedCore::get_cubes_dump() {
...
+// new function: in create_metrics you need the array of names
+std::vector<std::string> PolycubedCore::get_servicectrls_names_vector() {
+  std::vector<std::string> vector;
+  for (auto &it : servicectrls_map_) {
+    vector.push_back(it.first);
+  }
+  return vector;
+}
+
+// returns all the names of all cubes
```
Comment (frisso) [inline src/polycubed/src/polycubed_core.cpp:648]: All cubes? Or, all running cubes?

File: src/polycubed/src/rest_server.cpp [hunk 13]
```diff
@@ -303,6 +305,10 @@ void RestServer::setup_routes() {
   Hateoas::addRoute("topology", base);
 
   router_->addNotFoundHandler(bind(&RestServer::redirect, this));
+
+  // prometheus metrics
```
Comment (frisso) [inline src/polycubed/src/rest_server.cpp:309]: ==> // add new endpoint to retrieve Prometheus-compatible metrics

File: src/polycubed/src/rest_server.cpp [hunk 14]
```diff
@@ -931,5 +937,157 @@ const std::string &RestServer::getPort() {
     return port;
 }
 
+void RestServer::create_metrics() {
```
Comment (frisso) [inline src/polycubed/src/rest_server.cpp:940]: Write a small comment about what this function is about.

File: src/polycubed/src/rest_server.cpp [hunk 15]
```diff
@@ -931,5 +937,157 @@ const std::string &RestServer::getPort() {
...
+    registry = std::make_shared<prometheus::Registry>();
+    std::vector<std::string> name_services = core.get_servicectrls_names_vector();
+
+    // loop on all services (not cubes)
+    for (size_t i = 0; i < name_services.size(); i++) {
+      
+      // for every service I get the array of InfoMetric
+      auto metrics_service = core.get_service_controller(name_services[i]).get_infoMetrics();
+      for(auto metric: metrics_service)  {
+        if (metric.typeMetric == "C") {
```
Comment (frisso) [inline src/polycubed/src/rest_server.cpp:953]: What does it mean "C"? Shouldn't we replace with an enum/define?

File: src/polycubed/src/rest_server.cpp [hunk 16]
```diff
@@ -931,5 +937,157 @@ const std::string &RestServer::getPort() {
...
+    // collectables_ can collects al types of metrics (counter, gauge, histogram, summary)
+    collectables_.push_back(registry);
+  } catch (const std::runtime_error &e) {
+    logger->error("{0}", e.what());
+  }
+}
+
+
+// expose OpenMetrics
+void RestServer::get_metrics(const Pistache::Rest::Request &request,
```
Comment (frisso) [inline src/polycubed/src/rest_server.cpp:976]: Add a brief comment here.
It would be nice if you report also an example of the returned values.

File: src/polycubed/src/rest_server.cpp [hunk 17]
```diff
@@ -931,5 +937,157 @@ const std::string &RestServer::getPort() {
...
+        //every service has a map with the name of the metric and the InfoMetric struct (we need the pathmetric, typemetric, type-operation and a value)
+        auto mapInfoMetricsService = core.get_service_controller(serviceName).get_mapInfoMetrics();
+          // for every service
+              for(auto cubeName: core.get_service_controller(serviceName).get_names_cubes()) {
+                      std::string cubeStr = core.get_service_controller(serviceName).get_management_interface()->get_service()->ReadValue(cubeName, keys).message;
+                      jsoncons::json cubeJson = jsoncons::json::parse(cubeStr);
+
+                      for(auto& i: mapInfoMetricsService) {
+                       //TODO
+                       if(i.second.pathMetric.compare("TODO")!=0) {
```
Comment (frisso) [inline src/polycubed/src/rest_server.cpp:1040]: Add something more: what is this "todo"?
(no need to implement this code, but at least say what it should be supposed to do)

File: src/polycubed/src/server/Parser/YangNodes.tcpp [hunk 18]
```diff
@@ -294,13 +308,59 @@ void Yang::ParseList(
   parent->AddChild(std::move(resource));
 }
 
+/* TODO
```
Comment (frisso) [inline src/polycubed/src/server/Parser/YangNodes.tcpp:311]: What is this TODO (commented?)

File: src/polycubed/src/server/Parser/YangNodes.tcpp [hunk 19]
```diff
@@ -294,13 +308,59 @@ void Yang::ParseList(
...
+
 void Yang::ParseLeaf(
     const lys_node_leaf *leaf, std::queue<std::string> parsed_names,
     const std::shared_ptr<Resources::Body::ParentResource> &parent,
-    const bool generate_endpoint) const {
+    const bool generate_endpoint, ServiceMetadata *md) const {
   const auto &cli_example = FetchExtensionByIndex(
       leaf->ext,
       GetExtensionIndex(leaf->ext, leaf->ext_size, "cli-example", 11));
+  const auto &name_metric = FetchExtensionByIndex(
```
Comment (frisso) [inline src/polycubed/src/server/Parser/YangNodes.tcpp:345]: Add a comment here. Not clear what is this about.

File: src/services/pcn-ddosmitigator/datamodel/ddosmitigator.yang [hunk 20]
```diff
@@ -24,18 +24,30 @@ module ddosmitigator {
       type uint64;
       config false;
       description "Dropped Packets/s";
+      // polycube-base:name-metric "ddos_stats_pps_packets";
+      // polycube-base:type-metric "C";
```
Comment (frisso) [inline src/services/pcn-ddosmitigator/datamodel/ddosmitigator.yang:28]: Can we define some enum (e.g., COUNTER, etc)?

File: src/polycubed/src/rest_server.cpp (hunk 17 above)
Comment (pinoOgni) [reply src/polycubed/src/rest_server.cpp:1040]: Done :+1:

File: src/polycubed/src/rest_server.cpp (hunk 16 above)
Comment (pinoOgni) [reply src/polycubed/src/rest_server.cpp:976]: Added :+1:

File: src/polycubed/src/rest_server.cpp (hunk 14 above)
Comment (pinoOgni) [reply src/polycubed/src/rest_server.cpp:940]: Added :+1:  Sorry but some comments were in an old version of my code, thanks for the osservation.

File: src/libs/README.md (hunk 3 above)
Comment (pinoOgni) [reply src/libs/README.md:1]: Added :+1: thanks

File: src/polycubed/src/cube_factory_impl.cpp (hunk 11 above)
Comment (pinoOgni) [reply src/polycubed/src/cube_factory_impl.cpp:196]: Added :+1:

File: src/polycubed/src/polycubed_core.cpp (hunk 12 above)
Comment (pinoOgni) [reply src/polycubed/src/polycubed_core.cpp:648]: Yes, all running cubes. Changed.

File: src/polycubed/src/rest_server.cpp (hunk 13 above)
Comment (pinoOgni) [reply src/polycubed/src/rest_server.cpp:309]: Thanks! :+1:

File: src/polycubed/src/rest_server.cpp (hunk 17 above)
Comment (pinoOgni) [reply src/polycubed/src/rest_server.cpp:1040]: In reality it was not a classic TODO.  To clarify I added this comment:

I have considered the case where someone defines metrics in the yang of a service but for some reason does not define the value of the path-metric. Just write TODO and Polycube even having the data structures for that metric,  it will not try to retrieve the value.

File: src/libs/README.md (hunk 4 above)
Comment (pinoOgni) [reply src/libs/README.md:2]: Done  :+1:

File: src/libs/jsoncons/build_files/cmake/config.cmake.in (hunk 7 above)
Comment (pinoOgni) [reply src/libs/jsoncons/build_files/cmake/config.cmake.in:1]: Hi @frisso 
I tried to modify the CMakeLists as little as possible and I left this file because it is used to set 3 variables. I haven't investigated that much at the time, but now taking a look it seems like deleting this file and deleting it
```
configure_package_config_file (build_files / cmake / config.cmake.in
                              "$ {CMAKE_CURRENT_BINARY_DIR} / $ {} PROJECT_NAME Config.cmake"
                              INSTALL_DESTINATION $ {JSONCONS_CMAKECONFIG_INSTALL_DIR})

```

from CMakeLists.txt don't give any errors, but I don't know if the variables have remained set somewhere.

File: src/libs/jsoncons/build_files/cmake/Linux.cmake~ (hunk 6 above)
Comment (pinoOgni) [reply src/libs/jsoncons/build_files/cmake/Linux.cmake~:1]: No, it is automatically generated, thanks

File: src/polycubed/src/server/Parser/YangNodes.tcpp (hunk 18 above)
Comment (pinoOgni) [reply src/polycubed/src/server/Parser/YangNodes.tcpp:311]: Oops, I forgot to remove it, I wanted to create a parseAugment but I seem to remember that there were problems with the LibYang version and also, then I realize I didn't need to parse an Augment (until now).

File: src/polycubed/src/server/Parser/YangNodes.tcpp (hunk 19 above)
Comment (pinoOgni) [reply src/polycubed/src/server/Parser/YangNodes.tcpp:345]: Done :+1:

File: src/polycubed/src/CMakeLists.txt (hunk 10 above)
Comment (pinoOgni) [reply src/polycubed/src/CMakeLists.txt:81]: Done :+1:

File: src/libs/README.md (hunk 8 above)
Comment (pinoOgni) [reply src/libs/README.md:3]: Done :+1:

File: src/libs/README.md (hunk 5 above)
Comment (pinoOgni) [reply src/libs/README.md:9]: Done :+1:

File: src/services/pcn-ddosmitigator/datamodel/ddosmitigator.yang (hunk 20 above)
Comment (pinoOgni) [reply src/services/pcn-ddosmitigator/datamodel/ddosmitigator.yang:28]: Hi @frisso , I don't know if I understand the comment well though, but playing a bit with the Yang of services (to add metrics) and seeing some examples, I don't think it is possible to define an enumeration within an extension but you should create an enumeration in each node of each yang such as:
```
        leaf state {
          type enumeration {
            enum disabled;
            enum blocking;
            enum listening;
            enum learning;
            enum forwarding;
          }
          description "STP port state";
          config false;
        }
```
I'm not Yang's top expert, maybe there are other solutions but at the moment I haven't found any (for example typedef). 

Always if I understand correctly, a possible solution that I have not tried yet but that does not require touching Yang could be to put an enum "inside" the struct InfoMetric and modify the parser with a function that takes a string and returns an enum, for example "COUNTER" taken from the yang. 

About this comment https://github.com/polycube-network/polycube/pull/313/#discussion_r462302489, doing so perhaps it would be possible to eliminate
```
if (metric.typeMetric == "C")
```
If it is not strictly necessary I would soon do it in another PR.

Comment (pinoOgni) [reply src/services/pcn-ddosmitigator/datamodel/ddosmitigator.yang:28]: Hi @frisso , I don't know if I understand the comment well though, but playing a bit with the Yang of services (to add metrics) and seeing some examples, I don't think it is possible to define an enumeration within an extension (the metrics are defined as extensions in polycube-base) but you should create an enumeration in each node of each yang such as:
```
        leaf state {
          type enumeration {
            enum disabled;
            enum blocking;
            enum listening;
            enum learning;
            enum forwarding;
          }
          description "STP port state";
          config false;
        }
```
I'm not Yang's top expert, maybe there are other solutions but at the moment I haven't found any (for example typedef). 

Always if I understand correctly, a possible solution that I have not tried yet but that does not require touching Yang could be to put an enum "inside" the struct InfoMetric and modify the parser with a function that takes a string and returns an enum, for example "COUNTER" taken from the yang. 

If it is not strictly necessary I would soon do it in another PR.

File: src/polycubed/src/rest_server.cpp (hunk 15 above)
Comment (pinoOgni) [reply src/polycubed/src/rest_server.cpp:953]: I do not know if this is related to https://github.com/polycube-network/polycube/pull/313/#discussion_r462307552 but if only intends to define a #define or enum in the rest_server, yes, done :+1:

File: src/services/pcn-ddosmitigator/datamodel/ddosmitigator.yang (hunk 20 above)
Comment (frisso) [reply src/services/pcn-ddosmitigator/datamodel/ddosmitigator.yang]: Agreed, Giuseppe. Not strictly needed. Thanks!
